File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1:
            return head
     
        answer = ListNode()
        answer.next = head
        
        current = head
        tail = answer
        l = [current]
        while current.next is not None:
            current = current.next
            next_ = current.next
            l.append(current)
            if len(l) == k:
                for i in range(k-1):
                    l[(-1)*i - 1].next = l[(-1)*i - 2]
                l[0].next = next_
                tail.next = l[-1]
                tail = l[0]
                current = tail
                l.clear()
                

        return answer.next
            

                
           

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode newHead = new ListNode(0, head);
        ListNode prev = newHead;
        ListNode curr = head;
        int count = 1;

         while (true) {
            ListNode tail = prev;
            for (int i = 0; i < k; i++) {
                tail = tail.next;
                if (tail == null) {
                    return newHead.next;
                }
            }

            ListNode nextGroupHead = tail.next;
            ListNode[] reversedList = reverse(curr, tail);
            prev.next = reversedList[0]; // 새로 뒤집힌 헤드 연결
            reversedList[1].next = nextGroupHead; // 새로 뒤집힌 테일 연결
            
            prev = reversedList[1]; // prev를 업데이트
            curr = nextGroupHead; // curr을 다음 그룹 헤드로 업데이트
        }

    }

    ListNode[] reverse(ListNode head, ListNode tail) {
        ListNode prev = tail.next;
        ListNode curr = head;

        while (prev != tail) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return new ListNode[]{tail, head};
    }
}